pilot/pkg/serviceregistry/kube/controller/multicluster.go

```diff
@@ -51,107 +52,111 @@ type kubeController struct {
-		fetchCaRoot:           opts.FetchCaRoot,
-		caBundlePath:          opts.CABundlePath,
-		systemNamespace:       opts.SystemNamespace,
 		secretNamespace:       secretNamespace,
-		endpointMode:          opts.EndpointMode,
 		syncInterval:          opts.GetSyncInterval(),
 	}
 	mc.initSecretController(kc)
 
 	return mc, nil
 }
 
+func (m *Multicluster) EnableCertPatch(caBundlePath string, fetchCaRoot func() map[string]string) {
+	m.caBundlePath = caBundlePath
+	m.fetchCaRoot = fetchCaRoot
+}
+
 // AddMemberCluster is passed to the secret controller as a callback to be called
 // when a remote cluster is added.  This function needs to set up all the handlers
 // to watch for resources being added, deleted or changed on remote clusters.
-func (m *Multicluster) AddMemberCluster(clients kubelib.Client, clusterID string) error {
+func (m *Multicluster) AddMemberCluster(client kubelib.Client, clusterID string) error {
 	// stopCh to stop controller created here when cluster removed.
 	stopCh := make(chan struct{})
 	var remoteKubeController kubeController
 	remoteKubeController.stopCh = stopCh
 	m.m.Lock()
-	options := Options{
-		SystemNamespace:   m.systemNamespace,
-		WatchedNamespaces: m.WatchedNamespaces,
-		ResyncPeriod:      m.ResyncPeriod,
-		DomainSuffix:      m.DomainSuffix,
-		XDSUpdater:        m.XDSUpdater,
-		ClusterID:         clusterID,
-		NetworksWatcher:   m.networksWatcher,
-		Metrics:           m.metrics,
-		EndpointMode:      m.endpointMode,
-		SyncInterval:      m.syncInterval,
-	}
+	options := m.opts
+	options.ClusterID = clusterID
 	log.Infof("Initializing Kubernetes service registry %q", options.ClusterID)
-	kubectl := NewController(clients, options)
+	kubeRegistry := NewController(client, options)
 
-	remoteKubeController.Controller = kubectl
-	m.serviceController.AddRegistry(kubectl)
+	remoteKubeController.Controller = kubeRegistry
+	m.serviceController.AddRegistry(kubeRegistry)
 
 	m.remoteKubeControllers[clusterID] = &remoteKubeController
 	m.m.Unlock()
 
 	// Only need to add service handler for kubernetes registry as `initRegistryEventHandlers`,
 	// because when endpoints update `XDSUpdater.EDSUpdate` has already been called.
-	_ = kubectl.AppendServiceHandler(func(svc *model.Service, ev model.Event) { m.updateHandler(svc) })
+	_ = kubeRegistry.AppendServiceHandler(func(svc *model.Service, ev model.Event) { m.updateHandler(svc) })
+
+	// TODO move instance cache out of registries
+	if m.serviceEntryStore != nil && features.EnableServiceEntrySelectPods {
+		// Add an instance handler in the kubernetes registry to notify service entry store about pod events
+		_ = kubeRegistry.AppendWorkloadHandler(m.serviceEntryStore.WorkloadInstanceHandler)
+	}
+
+	if m.serviceEntryStore != nil && features.EnableK8SServiceSelectWorkloadEntries {
+		// Add an instance handler in the service entry store to notify kubernetes about workload entry events
+		_ = m.serviceEntryStore.AppendWorkloadHandler(kubeRegistry.WorkloadInstanceHandler)
```

Comment:
We support SE/WLE only in primary cluster now

Reply:
What is the reason? Shouldn't SE be able to select a Pod in any cluster?

Reply:
Yeah, the se should be able to select pods across clusters. What i mean is that we only watch SE/WLE from the primary cluster, so these handlers should be applied once rather than multi times. it may generate multiple duplicate service instances.

Reply:
We definitely need all the kube registries to have a handler so the SE store can read the Pods. 

Good point about all the kube controllers having the workload entry instances.. could cause dupes.

Reply:
Yeah, feel free to ping me when this is resolved